### PR TITLE
Support pushing to a sql-server instance

### PR DESCRIPTION
**Pull request by macneale4**
This change alters the remoteapi server behavior running in a sql-server to allow people to push to it, with some caveats:
 - User authenticates using the DOLT_REMOTE_PASSWORD environment vairable, and `--user` passed as a stored procedure or CLI flag
 - User has SuperUser privledges, due to the nature of the root update
 - The update of the server will be performed only if there are no working set changes and the commit being pushed is a fast forward.
 - The `--force` option will overwrite any changes on the target, including the working set changes

### CODE REVIEW

**Review by reltuk — APPROVED** [review]
A couple comments. In general it looks good and I think it's fine to roll forward with this.

My least favorite thing about this PR is the interface change in `Database`. I don't have a great suggestion for a near-at-hand solution. I tried out a couple of different ideas, but none of them looked all the good, especially with regards to still "supporting" `LD_1`. This PR has the following gaps in my opinion:

1) The Delete behavior is on `push origin :main` is still wrong. It doesn't update the remote workspace ref, and it doesn't require `-f` if that workspace ref exists and is different from the branch head.

2) The current approach does all the push work before it checks any of the invariants which will be enforced at Commit time. In general that seems worse then checking it up front and knowing that things will succeed unless there is concurrency while we're doing the push.

I also don't love the layering and architecture of the domain-specific rules being enforced in `Database` currently. I agree there's already a bit of that, most glaringly with regards to `FastForward`, but at least there it is in the domain model of `Commit` and the requested operation. This takes us pretty far afield of even the existing worst case, in my opinion, to where we have multi-head updates that are mediated by application business rules involving things like database root values, staged, working, merge state, etc.

All that being said, I don't think these should block rolling forward. We can come back and clean this up, and this code has the notable advantage of actually existing and being useful to users today.

File: go/libraries/doltcore/remotesrv/interceptors.go [hunk 1]
```diff
@@ -58,7 +61,9 @@ func (si *ServerInterceptor) Stream() grpc.StreamServerInterceptor {
 
 func (si *ServerInterceptor) Unary() grpc.UnaryServerInterceptor {
 	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
-		if err := si.authenticate(ctx); err != nil {
+		_, needSuperUser := req.(*remotesapi.CommitRequest)
```
[the remark this answers is not in the document]
Comment (macneale4) [reply go/libraries/doltcore/remotesrv/interceptors.go]: made the list for all explicit, and rejected unexpected methods

**Comment by reltuk** [comment]
A thought crossed my mind when investigating a bug in a different context.

Currently the SQL engine's implementation of the interface that `remotesrv` uses to get a database by name will automatically create the requested database if it does not exist. See https://github.com/dolthub/dolt/blob/main/go/libraries/doltcore/sqle/remotesrv.go#L35

We need to decide as part of this PR:

1) If we want to people to be able to create new databases by pushing a branch to a novel database name.
2) If we don't, we need to disable that functionality for the sql-server context. (It needs to stick around for the cluster replication context).
3) Regardless of what we decide, we should add a test for it.

**Comment by macneale4** [comment]
> Regardless of what we decide, we should add a test for it.

Disabled this behavior and added a test.

**Pull request merged**